staging/src/k8s.io/legacy-cloud-providers/gce/gce_loadbalancer_metrics.go

```diff
@@ -0,0 +1,165 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package gce
+
+import (
+	"sync"
+	"time"
+
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/component-base/metrics"
+	"k8s.io/component-base/metrics/legacyregistry"
+	"k8s.io/klog/v2"
+)
+
+const (
+	label = "feature"
+)
+
+var (
+	metricsInterval = 10 * time.Minute
+	l4ILBCount = metrics.NewGaugeVec(
+		&metrics.GaugeOpts{
+			Name: "number_of_l4_ilbs",
+			Help: "Number of L4 ILBs",
+		},
+		[]string{label},
+	)
+)
+
+// init registers L4 internal loadbalancer usage metrics.
+func init() {
+	klog.V(3).Infof("Registering Service Controller usage metrics %v", l4ILBCount)
+	legacyregistry.MustRegister(l4ILBCount)
+}
+
+// LoadBalancerMetrics is a cache that contains loadbalancer service resource
+// states for computing usage metrics.
+type LoadBalancerMetrics struct {
+	// l4ILBServiceMap is a map of service key and L4 ILB service state.
+	l4ILBServiceMap map[string]L4ILBServiceState
+
+	sync.Mutex
+}
+
+type feature string
+
+func (f feature) String() string {
+	return string(f)
+}
+
+const(
+	l4ILBService      = feature("L4ILBService")
+	l4ILBGlobalAccess = feature("L4ILBGlobalAccess")
+	l4ILBCustomSubnet = feature("L4ILBCustomSubnet")
+)
+
+// L4ILBServiceState contains Internal Loadbalancer feature states as specified
+// in k8s Service.
+type L4ILBServiceState struct {
+	// EnabledGlobalAccess specifies if Global Access is enabled.
+	EnabledGlobalAccess bool
+	// EnabledCustomSubNet specifies if Custom Subnet is enabled.
+	EnabledCustomSubnet bool
+}
+
+// loadbalancerMetricsCollector is an interface to update/delete L4 loadbalancer
+// states in the cache that is used for computing L4 Loadbalancer usage metrics.
+type loadbalancerMetricsCollector interface {
+	// Run starts a goroutine to compute and export metrics a periodic interval.
+	Run(stopCh <-chan struct{})
+	// SetL4ILBService adds/updates L4 ILB service state for given service key.
+	SetL4ILBService(svcKey string, state L4ILBServiceState)
+	// DeleteL4ILBService removes the given L4 ILB service key.
+	DeleteL4ILBService(svcKey string)
+}
+
+// newLoadBalancerMetrics initializes LoadBalancerMetrics and starts a goroutine
+// to compute and export metrics periodically.
+func newLoadBalancerMetrics() loadbalancerMetricsCollector {
+	return &LoadBalancerMetrics{
+		l4ILBServiceMap: make(map[string]L4ILBServiceState),
+	}
+}
+
+// Run implements loadbalancerMetricsCollector.
+func (lm *LoadBalancerMetrics) Run(stopCh <-chan struct{}) {
+	klog.V(3).Infof("Loadbalancer Metrics initialized. Metrics will be exported at an interval of %v", metricsInterval)
+	// Compute and export metrics periodically.
+	go func() {
+		// Wait for ingress states to be populated in the cache before computing metrics.
+		time.Sleep(metricsInterval)
```

Comment:
Do you need time.Sleep? Looks like wait.Until will block anyway until stopped?

Reply:
This is to ensure that the metrics are not exported immediately on controller start up. Maybe we would want to wait for some initial time (in this case 10 minutes) for all the services to sync, before exporting the metrics.

Reply:
ok, makes sense.